### PR TITLE
feat: Support for Clipboard on WPF and Gtk

### PR DESCRIPTION
**Pull request by trungnt2910**
GitHub Issue (If applicable): a part of #6179

## PR Type

What kind of change does this PR introduce?

- Feature

## What is the current behavior?
Clipboard does nothing on Skia.Wpf and Skia.Gtk targets.

## What is the new behavior?
Clipboard should copy data.

Some more details:
- `ContentChanged` event: Invokes properly, however, on both platforms, the functions may be invoke multiple times as many data targets are set.
- Text, Bitmap, and Html data: Works properly on both platforms.
- Rtf data: Works properly on WPF, can be detected on Linux if somehow Rtf data is set (for example when sharing clipboard using VMWare), although Rtf data is not natively found on Linux.
- Files: Works properly on WPF, should (theoretically) work on Gtk, however does ~~**not** work on popular distros using Nautilus (which stores copied files as _text_)~~. GNOME 40 [fixed the bug](https://gitlab.gnome.org/GNOME/nautilus/-/issues/634), I've tested it on Ubuntu and files are now working fine.
- Uri: 
   + There seems to be no Uri content on WPF, so Uri is not supported at all.
   + On Gtk, there are methods to obtain Uris, however, there is no official way to set an Uri to clipboard.

## PR Checklist

Please check if your PR fulfills the following requirements:

- [ ] Docs have been added/updated which fit [documentation template](https://github.com/unoplatform/uno/blob/master/doc/.feature-template.md) (for bug fixes / features)
- [ ] [Unit Tests and/or UI Tests](https://github.com/unoplatform/uno/blob/master/doc/articles/uno-development/working-with-the-samples-apps.md) for the changes have been added (for bug fixes / features) (if applicable)
- [ ] Validated PR `Screenshots Compare Test Run` results.
- [x] Contains **NO** breaking changes
- [ ] Associated with an issue (GitHub or internal) and uses the [automatic close keywords](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue).
- [x] Commits must be following the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/#summary) specification.

## Other information

Internal Issue (If applicable):

### CODE REVIEW

File: src/Uno.UI.Runtime.Skia.Wpf/Extensions/ApplicationModel/DataTransfer/ClipboardExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,231 @@
...
+
+			Clipboard.SetDataObject(wpfData);
+		}
+
+		private IntPtr OnWmMessage(IntPtr hwnd, int msg, IntPtr wparamOriginal, IntPtr lparamOriginal, ref bool handled)
+		{
+			switch (msg)
+			{
+				case WM_CLIPBOARDUPDATE:
+                {
```
Comment (MartinZikmund) [inline src/Uno.UI.Runtime.Skia.Wpf/Extensions/ApplicationModel/DataTransfer/ClipboardExtensions.cs:221]: Seems indentation is wrong here

File: src/Uno.UI.Runtime.Skia.Wpf/Extensions/ApplicationModel/DataTransfer/ClipboardExtensions.cs [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+
+		public void StartContentChanged()
+		{
+			if (_hwndSource != null)
+			{
+				_hwndSource.AddHook(OnWmMessage);
+				ClipboardNativeFunctions.AddClipboardFormatListener(_hwndSource.Handle);
+			}
+			else
+            {
```
Comment (MartinZikmund) [inline src/Uno.UI.Runtime.Skia.Wpf/Extensions/ApplicationModel/DataTransfer/ClipboardExtensions.cs:69]: And here the indentation seems also distorted 😀

**Review by MartinZikmund — APPROVED** [review]
Awesome!

File: src/Uno.UI.Runtime.Skia.Gtk/ApplicationModel/DataTransfer/ClipboardExtensions.cs [hunk 3]
```diff
@@ -0,0 +1,336 @@
...
+				{
+					var bytes = Encoding.UTF8.GetBytes(new[] { ch });
+					foreach (var b in bytes)
+					{
+						uri.Append($"%{b:X2}");
+					}
+				}
+			}
+			return "file://" + uri;
+		}
```
Comment (carldebilly) [inline src/Uno.UI.Runtime.Skia.Gtk/ApplicationModel/DataTransfer/ClipboardExtensions.cs:334]: Why you're not using [`Uri.EscapeDataString`](https://docs.microsoft.com/en-us/dotnet/api/system.uri.escapedatastring)?

File: src/Uno.UI.Runtime.Skia.Wpf/Extensions/ApplicationModel/DataTransfer/ClipboardExtensions.cs (hunk 1 above)
Comment (trungnt2910) [reply src/Uno.UI.Runtime.Skia.Wpf/Extensions/ApplicationModel/DataTransfer/ClipboardExtensions.cs:221]: :trollface:

File: src/Uno.UI.Runtime.Skia.Wpf/Extensions/ApplicationModel/DataTransfer/ClipboardExtensions.cs (hunk 2 above)
Comment (trungnt2910) [reply src/Uno.UI.Runtime.Skia.Wpf/Extensions/ApplicationModel/DataTransfer/ClipboardExtensions.cs:69]: Probably the consequences of copy-paste from my experimental app to the library. Sorry for that.
(And why doesn't Codacy catch it?)

File: src/Uno.UI.Runtime.Skia.Gtk/ApplicationModel/DataTransfer/ClipboardExtensions.cs (hunk 3 above)
Comment (trungnt2910) [reply src/Uno.UI.Runtime.Skia.Gtk/ApplicationModel/DataTransfer/ClipboardExtensions.cs:334]: https://dotnetfiddle.net/wO60Dv
It doesn't handle the DirectorySeparatorChar properly.

Comment (trungnt2910) [reply src/Uno.UI.Runtime.Skia.Gtk/ApplicationModel/DataTransfer/ClipboardExtensions.cs:334]: We might use `Uri.EscapeDataString(str).Replace(Uri.EscapeDataString(Path.DirectorySeparatorChar), Path.DirectorySeparatorChar)`, which is a lot slower but uses the library. Should I change to that?

**Comment by MartinZikmund** [comment]
@trungnt2910 can you resolve conflicts on this one so we can merge it?

**Comment by trungnt2910** [comment]
Yeah, I'll fix this now.

File: src/Uno.UI.Runtime.Skia.Gtk/ApplicationModel/DataTransfer/ClipboardExtensions.cs (hunk 3 above)
Comment (MartinZikmund) [reply src/Uno.UI.Runtime.Skia.Gtk/ApplicationModel/DataTransfer/ClipboardExtensions.cs:334]: @carldebilly can this be resolved?

Comment (jeromelaban) [reply src/Uno.UI.Runtime.Skia.Gtk/ApplicationModel/DataTransfer/ClipboardExtensions.cs:334]: `new Uri("").ToString()` does what you need: https://dotnetfiddle.net/5qJNKS.

You still need to replace `\` to `/` before, though.

Comment (trungnt2910) [reply src/Uno.UI.Runtime.Skia.Gtk/ApplicationModel/DataTransfer/ClipboardExtensions.cs:334]: @jeromelaban No, it doesn't, it's apparent in your example.

For example, for the file `/home/user/%51.txt/`, which is a perfectly valid path, to a file named `%51.txt`, the function returns `file:///home/user/Q.txt/`, which points to another file, named `Q.txt`.

The local path also returns `/home/user/Q.txt/`, which is incorrect.

This is actually the first solution I think of, but it doesn't work in this edge case.

Comment (jeromelaban) [reply src/Uno.UI.Runtime.Skia.Gtk/ApplicationModel/DataTransfer/ClipboardExtensions.cs:334]: Got it. We'll still need to unit test this, particularly for those edge cases.

Can you move this method to a class next to Clipboard.skia.cs (e.g. FileUriHelper) and add a test here: https://github.com/unoplatform/uno/tree/master/src/Uno.UI.Tests ?

DataRows tests will be useful, similar to this one:
https://github.com/unoplatform/uno/blob/d57d8ed572184968a801dda815ac30bba96122cc/src/Uno.UI.Tests/Windows_Globalization/Given_CalendarFormatter.cs#L13-L20

**Review by jeromelaban — APPROVED** [review]

File: src/Uno.UI.Runtime.Skia.Gtk/ApplicationModel/DataTransfer/ClipboardExtensions.cs (hunk 3 above)
Comment (trungnt2910) [reply src/Uno.UI.Runtime.Skia.Gtk/ApplicationModel/DataTransfer/ClipboardExtensions.cs]: I've added the test, so I think this should be resolved for now?
